### PR TITLE
Arithmetical instrs

### PR DESCRIPTION
**Pull request by includealex**

### CODE REVIEW

File: src/Executor.cpp [hunk 1]
```diff
@@ -278,51 +278,71 @@ void ExecutorADDI(Hart* hart, const DecodedInstruction& instr) {
 void ExecutorSLLI(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("slli    x%d, x%d, %d\n", instr.rd, instr.rs1, instr.shamt);
 
+    uint64_t res = (hart->getReg(instr.rs1)) << (instr.imm);
```
Comment (OAMichael) [inline src/Executor.cpp:281]: Maybe use instr.shamt instead of instr.imm?

File: src/Executor.cpp [hunk 2]
```diff
@@ -278,51 +278,71 @@ void ExecutorADDI(Hart* hart, const DecodedInstruction& instr) {
...
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorSLTI(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("slti    x%d, x%d, x%d\n", instr.rd, instr.rs1, instr.rs2);
 
+    auto rs1_signed = sext(hart->getReg(instr.rs1), 0);
```
Comment (OAMichael) [inline src/Executor.cpp:289]: I believe you must sign extend imm, not rs1

File: src/Executor.cpp [hunk 3]
```diff
@@ -278,51 +278,71 @@ void ExecutorADDI(Hart* hart, const DecodedInstruction& instr) {
...
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorSLTIU(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("sltiu   x%d, x%d, x%d\n", instr.rd, instr.rs1, instr.rs2);
 
+    auto left = hart->getReg(instr.rs1);
+    auto right = sext(instr.imm, instr.immSignBitNum);
+    uint64_t res = left < right ? 1 : 0;
```
Comment (OAMichael) [inline src/Executor.cpp:300]: Check signed and unsigned comparisons

File: src/Executor.cpp [hunk 4]
```diff
@@ -278,51 +278,71 @@ void ExecutorADDI(Hart* hart, const DecodedInstruction& instr) {
...
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorXORI(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION(
         "xori    x%d, x%d, %ld\n", instr.rd, instr.rs1, sext(instr.imm, instr.immSignBitNum));
 
+    uint64_t res = (hart->getReg(instr.rs1)) ^ instr.imm;
```
Comment (OAMichael) [inline src/Executor.cpp:309]: There you have to sign extend immediate

File: src/Executor.cpp [hunk 5]
```diff
@@ -278,51 +278,71 @@ void ExecutorADDI(Hart* hart, const DecodedInstruction& instr) {
...
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorSRLI(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("srli    x%d, x%d, %d\n", instr.rd, instr.rs1, instr.shamt);
 
+    auto res = (hart->getReg(instr.rs1)) >> (instr.imm);
```
Comment (OAMichael) [inline src/Executor.cpp:317]: Use shamt instead of imm

File: src/Executor.cpp [hunk 6]
```diff
@@ -278,51 +278,71 @@ void ExecutorADDI(Hart* hart, const DecodedInstruction& instr) {
...
+    auto res = rs1_signed >> (instr.imm);
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorORI(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION(
         "ori     x%d, x%d, %ld\n", instr.rd, instr.rs1, sext(instr.imm, instr.immSignBitNum));
 
+    uint64_t res = (hart->getReg(instr.rs1)) | instr.imm;
```
Comment (OAMichael) [inline src/Executor.cpp:335]: You have to sign extend immediate

File: src/Executor.cpp [hunk 7]
```diff
@@ -278,51 +278,71 @@ void ExecutorADDI(Hart* hart, const DecodedInstruction& instr) {
...
 
+    auto res = (hart->getReg(instr.rs1)) >> (instr.imm);
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorSRAI(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("srai    x%d, x%d, %d\n", instr.rd, instr.rs1, instr.shamt);
 
+    auto rs1_signed = sext(hart->getReg(instr.rs1), 0);
```
Comment (OAMichael) [inline src/Executor.cpp:325]: Check this. This is incorrect

File: src/Executor.cpp [hunk 8]
```diff
@@ -278,51 +278,71 @@ void ExecutorADDI(Hart* hart, const DecodedInstruction& instr) {
...
+    uint64_t res = (hart->getReg(instr.rs1)) | instr.imm;
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorANDI(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION(
         "andi    x%d, x%d, %ld\n", instr.rd, instr.rs1, sext(instr.imm, instr.immSignBitNum));
 
+    uint64_t res = (hart->getReg(instr.rs1)) & instr.imm;
```
Comment (OAMichael) [inline src/Executor.cpp:344]: You have to sign extend immediate

File: src/Executor.cpp [hunk 9]
```diff
@@ -334,25 +354,36 @@ void ExecutorADDIW(Hart* hart, const DecodedInstruction& instr) {
...
 
     hart->setReg(instr.rd, sext(sum32, 31));
-
     hart->incrementPC();
 }
 
 void ExecutorSLLIW(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("slliw   x%d, x%d, %d\n", instr.rd, instr.rs1, instr.shamt);
 
+    uint64_t res = (hart->getReg(instr.rs1)) << instr.imm;
```
Comment (OAMichael) [inline src/Executor.cpp:363]: Use shamt instead of imm

File: src/Executor.cpp [hunk 10]
```diff
@@ -334,25 +354,36 @@ void ExecutorADDIW(Hart* hart, const DecodedInstruction& instr) {
...
+    uint32_t res32 = res;
+
+    hart->setReg(instr.rd, sext(res32, 31));
     hart->incrementPC();
 }
 
 void ExecutorSRLIW(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("srliw   x%d, x%d, %d\n", instr.rd, instr.rs1, instr.shamt);
 
+    uint64_t res = (hart->getReg(instr.rs1)) >> instr.imm;
```
Comment (OAMichael) [inline src/Executor.cpp:373]: Use shamt instead of imm

File: src/Executor.cpp [hunk 11]
```diff
@@ -334,25 +354,36 @@ void ExecutorADDIW(Hart* hart, const DecodedInstruction& instr) {
...
+    uint32_t res32 = res;
+
+    hart->setReg(instr.rd, sext(res32, 31));
     hart->incrementPC();
 }
 
 void ExecutorSRAIW(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("sraiw   x%d, x%d, %d\n", instr.rd, instr.rs1, instr.shamt);
 
+    int32_t rs1_signed32 = sext(hart->getReg(instr.rs1), 0);
```
Comment (OAMichael) [inline src/Executor.cpp:383]: Check this. This is incorrect

File: src/Executor.cpp [hunk 12]
```diff
@@ -361,60 +392,82 @@ void ExecutorSRAIW(Hart* hart, const DecodedInstruction& instr) {
 void ExecutorADD(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("add     x%d, x%d, x%d\n", instr.rd, instr.rs1, instr.rs2);
 
+    uint64_t res = sext(hart->getReg(instr.rs1), 0) + sext(hart->getReg(instr.rs2), 0);
```
Comment (OAMichael) [inline src/Executor.cpp:395]: You do not have to sign extend register values

File: src/Executor.cpp [hunk 13]
```diff
@@ -361,60 +392,82 @@ void ExecutorSRAIW(Hart* hart, const DecodedInstruction& instr) {
...
+    auto res = hart->getReg(instr.rs1) << hart->getReg(instr.rs2);
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorSLT(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("slt     x%d, x%d, x%d\n", instr.rd, instr.rs1, instr.rs2);
 
+    auto left = sext(hart->getReg(instr.rs1), 0);
+    auto right = sext(hart->getReg(instr.rs2), 0);
```
Comment (OAMichael) [inline src/Executor.cpp:412]: Mind correct signed and unsigned comparisons

File: src/Executor.cpp [hunk 14]
```diff
@@ -361,60 +392,82 @@ void ExecutorSRAIW(Hart* hart, const DecodedInstruction& instr) {
...
 
+    uint64_t res = (hart->getReg(instr.rs1)) & (hart->getReg(instr.rs2));
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorSUB(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("sub     x%d, x%d, x%d\n", instr.rd, instr.rs1, instr.rs2);
 
+    uint64_t res = sext(hart->getReg(instr.rs1), 0) - sext(hart->getReg(instr.rs2), 0);
```
Comment (OAMichael) [inline src/Executor.cpp:461]: Do you need to somehow transform register values?

File: src/Executor.cpp [hunk 15]
```diff
@@ -361,60 +392,82 @@ void ExecutorSRAIW(Hart* hart, const DecodedInstruction& instr) {
...
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorSRA(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("sra     x%d, x%d, x%d\n", instr.rd, instr.rs1, instr.rs2);
 
+    uint64_t res = (sext(hart->getReg(instr.rs1), 0)) >> hart->getReg(instr.rs2);
```
Comment (OAMichael) [inline src/Executor.cpp:469]: Just try to cast to int64_t

File: src/Executor.cpp [hunk 16]
```diff
@@ -423,33 +476,43 @@ void ExecutorADDW(Hart* hart, const DecodedInstruction& instr) {
...
+    uint64_t res = (hart->getReg(instr.rs1)) >> (hart->getReg(instr.rs2));
+    uint32_t res32 = res;
+    hart->setReg(instr.rd, sext(res32, 31));
     hart->incrementPC();
 }
 
 void ExecutorSRAW(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("sraw    x%d, x%d, x%d\n", instr.rd, instr.rs1, instr.rs2);
 
+    uint64_t res = (sext(hart->getReg(instr.rs1), 0)) >> (hart->getReg(instr.rs2));
```
Comment (OAMichael) [inline src/Executor.cpp:513]: Just try to cast to int64_t

**Comment by includealex** [comment]
@OAMichael please check

File: src/Executor.cpp [hunk 17]
```diff
@@ -278,51 +278,69 @@ void ExecutorADDI(Hart* hart, const DecodedInstruction& instr) {
...
 
+    uint64_t res = hart->getReg(instr.rs1) << (instr.shamt);
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorSLTI(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("slti    x%d, x%d, x%d\n", instr.rd, instr.rs1, instr.rs2);
 
+    uint64_t res = hart->getReg(instr.rs1) < instr.imm;
```
Comment (OAMichael) [inline src/Executor.cpp:289]: You need to sign extend immediate first. Here is sample from spec:

> SLTI (set less than immediate) places the value 1 in register rd if register rs1 is less than the signextended immediate when both are treated as signed numbers, else 0 is written to rd.

File: src/Executor.cpp [hunk 18]
```diff
@@ -278,51 +278,69 @@ void ExecutorADDI(Hart* hart, const DecodedInstruction& instr) {
...
+    hart->setReg(instr.rd, res);
     hart->incrementPC();
 }
 
 void ExecutorSLTIU(Hart* hart, const DecodedInstruction& instr) {
     DEBUG_INSTRUCTION("sltiu   x%d, x%d, x%d\n", instr.rd, instr.rs1, instr.rs2);
 
+    auto left = hart->getReg(instr.rs1);
```
Comment (OAMichael) [inline src/Executor.cpp:297]: Use explicit types instead of auto here. You need exactly `uint64_t`. And in SLTI you need do the same but `int64_t`

**Comment by panferovi** [comment]
OK

**Comment by includealex** [comment]
@panferovi @OAMichael PTL

**Comment by panferovi** [comment]
Ok

**Review by panferovi — APPROVED** [review]

**Review by OAMichael — APPROVED** [review]
OK

**Pull request merged**